src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,12 +394,156 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
                     }
                     else
                     {
-                        ApplyResourceInNestedProperty(edmProperty, resource, resourceWrapper, readContext);
+                        ODataResourceWrapper resWrapper = resourceWrapper;
+                        if (resourceWrapper.Resource.Id != null)
+                        {
+                            resWrapper = UpdateResourceWrapper(edmProperty.Type, resWrapper.Resource.Id, readContext);
+
+                        }
+                        ApplyResourceInNestedProperty(edmProperty, resource, resWrapper, readContext);
                     }
                 }
             }
         }
 
+
+        private ODataResourceSetWrapper CreateResourceSetWrapper(IEdmCollectionTypeReference edmPropertyType,
+    IList<ODataEntityReferenceLinkWrapper> refLinks, ODataDeserializerContext readContext)
+        {
+            ODataResourceSet resourceSet = new ODataResourceSet
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            IEdmTypeReference elementType = edmPropertyType.ElementType();
+            ODataResourceSetWrapper resourceSetWrapper = new ODataResourceSetWrapper(resourceSet);
+            foreach (ODataEntityReferenceLinkWrapper refLinkWrapper in refLinks)
+            {
+                ODataResourceWrapper resourceWrapper = CreateResourceWrapper(elementType, refLinkWrapper, readContext);
+                resourceSetWrapper.Resources.Add(resourceWrapper);
+            }
+
+            return resourceSetWrapper;
+        }
+
+        private ODataResourceWrapper CreateResourceWrapper(IEdmTypeReference edmPropertyType, ODataEntityReferenceLinkWrapper refLink, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(readContext != null);
+
+            ODataResource resource = new ODataResource
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            Uri serviceRootUri = null;
+            string serviceRoot = "";
+            DefaultODataPathHandler pathHandler = new DefaultODataPathHandler();
+
+            IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+            IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+            serviceRoot = readContext.InternalUrlHelper.CreateODataLink(
+                readContext.InternalRequest.Context.RouteName,
+                readContext.InternalRequest.PathHandler,
+                new List<ODataPathSegment>());
+
+            serviceRootUri = new Uri(serviceRoot, UriKind.Absolute);
+
+            var request = readContext.Request;
+            IEdmModel model = readContext.Model;
+
+            ODataPath odataPath = pathHandler.Parse(serviceRoot, refLink.EntityReferenceLink.Url.OriginalString,readContext.InternalRequest.RequestContainer);
+            IList<KeySegment> keySegments = odataPath.Segments.OfType<KeySegment>().ToList();
+            if (keySegments == null)
```

Comment:
Firstly, I don't think `odataPath.Segments.OfType<KeySegment>().ToList()` will ever return a `null`. It'll be an empty list at worst.

Secondly, if we find no key segments, should we return null or throw an exception? To find ourselves here, I'd think that the reference to the related entity was most probably invalid - I could be wrong, plus I'm not sure we'd get this far if it was..., but if it was wrong, we can't successfully create a relationship between the entity targeted for creation and the related entity. According to the [spec](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_LinktoRelatedEntitiesWhenCreatinganE), _the service MUST never create an entity in a partially valid state (with the navigation property unset)_